### ISSUE TITLE
dns resolve issue with link

### ISSUE THREAD

**Issue by osallou** [rancher/rancher#3987]
Rancher Version:
0.59.0
Docker Version:
1.10.1
OS: 
CentOS Linux release 7.1.1503
Steps to Reproduce: 
After a reboot, service links do not resolve anymore.
Launched in a stack some services (symfonytest, ...). Then I launch an other service in the same stack with a link to one of previous containers, I use a link because I "rename" the linked container :  link symfonytest=> php
Results: 
when new service is started if failed, "php" name is not resolved.

Labels and env on new services are:

            "Env": [
                "SYMFONYTEST_1_PORT_9000_TCP=tcp://php:9000",
                "SYMFONYTEST_1_NAME=/cattle/symfonytest_1",
                "SYMFONYPROJECTTEST_SYMFONYTEST_1_PORT_9000_TCP_PORT=9000",
                "PHP_NAME=/cattle/php",
                "SYMFONYPROJECTTEST_SYMFONYTEST_1_PORT=tcp://php:9000",
                "SYMFONYTEST_1_PORT_9000_TCP_PORT=9000",
                "PHP_PORT_9000_TCP_ADDR=php",
                "PHP_PORT_9000_TCP=tcp://php:9000",
                "SYMFONYPROJECTTEST_SYMFONYTEST_1_PORT_9000_TCP_PROTO=tcp",
                "SYMFONYTEST_1_PORT_9000_TCP_PROTO=tcp",
                "SYMFONYPROJECTTEST_SYMFONYTEST_1_NAME=/cattle/SymfonyProjectTest_symfonytest_1",
                "SYMFONYPROJECTTEST_SYMFONYTEST_1_PORT_9000_TCP=tcp://php:9000",
                "PHP_PORT_9000_TCP_PROTO=tcp",
                "PHP_PORT_9000_TCP_PORT=9000",
                "SYMFONYPROJECTTEST_SYMFONYTEST_1_PORT_9000_TCP_ADDR=php",
                "SYMFONYTEST_1_PORT=tcp://php:9000",
                "PHP_PORT=tcp://php:9000",
                "SYMFONYTEST_1_PORT_9000_TCP_ADDR=php",
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
            ],

            "Labels": {
                "io.rancher.container.ip": "10.42.199.4/16",
                "io.rancher.container.pull_image": "always",
                "io.rancher.container.start_once": "true",
                "io.rancher.container.uuid": "ec7101ef-b9e9-4cad-b4b6-18fd17c8072a",
                "io.rancher.project.name": "SymfonyProjectTest",
                "io.rancher.project_service.name": "SymfonyProjectTest/symfonynginx",
                "io.rancher.scheduler.affinity:container_label_soft": "io.rancher.service.deployment.unit=3a346e22-97d5-4f87-8263-5ae2a2303678",
                "io.rancher.service.deployment.unit": "3a346e22-97d5-4f87-8263-5ae2a2303678",
                "io.rancher.service.launch.config": "io.rancher.service.primary.launch.config",
                "io.rancher.stack.name": "SymfonyProjectTest",
                "io.rancher.stack_service.name": "SymfonyProjectTest/symfonynginx"
            }

Expected:
should resolve linked name

**Comment by deniseschannon** [comment]
Can you share the docker-compose.yml of the stack?

**Comment by cloudnautique** [comment]
Also, does it try to resolve immediately? There is a slight lag between when the container comes up and the DNS resolver is able to answer. It can range from  1<n<10 seconds depending on load.

**Comment by osallou** [comment]
Regarding resolution , i start an nginx that connect to the *php* service.
It raised an unresolved php name error at startup.
It used to work however, my concerns occured only after a derver reboot and
restarting rancher containers. I will send compose when i get my computer.

Le jeu. 17 mars 2016 06:27, Bill Maxwell <notifications@github.com> a
écrit :

**Comment by osallou** [comment]
Here is compose file

    symfonytest:
      labels:
        io.rancher.container.pull_image: always
      image: osallou/symfony-php
      volumes_from:
      - 16086b96-eb72-4340-92c9-b4c506d6e9b9
    symfonynginx:
      ports:
      - 8000:80/tcp
      labels:
        io.rancher.container.pull_image: always
        io.rancher.container.start_once: 'true'
      image: osallou/symfony-nginx
      links:
      - symfonytest:php
      volumes:
      - /root/symfony-tests/project1-logs:/var/log/nginx
      volumes_from:
      - 16086b96-eb72-4340-92c9-b4c506d6e9b9
    code:
      labels:
        io.rancher.container.pull_image: always
      tty: true
      image: osallou/symfony-code
      volumes:
      - /root/symfony-tests/project1:/var/www/symfony
      - /root/symfony-tests/project1-logs:/var/www/symfony/app/logs
      stdin_open: true

**Comment by alena1108** [comment]
@vincent99 need your help on that. Deployed the stack using compose file of @osallou, ran tcpdump on network agent while bringing up container "symfonynginx" (10.42.14.151) that tries to resolve "php. (alias for "symfonytest", ip address 10.42.9.202). Here is the tcpdump: https://gist.github.com/alena1108/8c3ab500048457432eae. Note that the first request is coming from "172.17.0.5" - ip address specified in /etc/hosts on client ""symfonynginx"). Eventually I do see php. address being returned here: https://gist.github.com/alena1108/8c3ab500048457432eae#file-gistfile1-txt-L63

Not sure if there is anything obviously wrong that tcpdump reveals.

When I ssh in the container and try to ping "php", it resolves fine.

**Comment by osallou** [comment]
As someone said in a previous comment, resolution may take a little time.
If dns time resolution is the issue, it is a problem for any service trying
to connect at startup to an other one.
After reboot, i did not check ntp, and will check asap, could it have some
influence?

Le ven. 18 mars 2016 01:33, Alena Prokharchyk <notifications@github.com> a
écrit :

**Comment by vincent99** [comment]
So there's 2 problems here..

1) The container code is running before the answers and/or resolv.conf are updated.  So the name fails to resolve when nginx starts.  We should have this fixed before GA but you could address this on your side with an entrypoint that waits for the name to resolve before starting nginx.

2) Assuming it did resolve, this nginx config will not really work right:
```
upstream php-upstream { server php:9000; }
```

This will cause nginx to resolve `php` exactly once on startup and never again (until next startup).  So when the addresses of `php` change because containers come and go, it will still have stale addresses in memory and send traffic to bad targets.  To address this you need to configure a resolver in nginx and use it, with something like:

```
    location / {
      resolver 169.254.169.250 valid=5s ipv6=off;
      set $php "php";
      proxy_pass http://$php:8000;
    }
```

**Comment by osallou** [comment]
I confirm that adding a *sleep 10* before nginx startup solves the issue

**Comment by ibuildthecloud** [comment]
Changes ~~shear madness~~ have been done such that DNS/Metadata should be available when the container start.  I tested by deploying this stack in Rancher https://github.com/ibuildthecloud/metadata-ping/blob/master/package/docker-compose.yml which tests the DNS and metadata right at start.

**Comment by LLParse** [comment]
Using my [test catalog entry](https://github.com/LLParse/community-catalog/tree/issue-3987/templates/issue-3987), I am able to reproduce the issue in versions <=1.0.0-r2 and verify the issue is fixed in 1.0.0-rc3.
